lib/src/widgets/shape.dart

```diff
@@ -3,50 +3,90 @@ import 'package:flutter/widgets.dart';
 
 import '../models.dart';
 
-class ArrowWidget extends StatelessWidget {
-  const ArrowWidget({
+class ShapeWidget extends StatelessWidget {
+  const ShapeWidget({
     super.key,
-    required this.color,
     required this.size,
     required this.orientation,
-    required this.fromCoord,
-    required this.toCoord,
+    required this.shape,
+    this.widthRatio = 1 / 6.4, // Default pulled from lichess.org
   });
 
-  final Color color;
   final double size;
   final Side orientation;
-  final Coord fromCoord;
-  final Coord toCoord;
+  final Shape shape;
+  final double widthRatio; // Only applies to arrows
 
   @override
   Widget build(BuildContext context) {
     return SizedBox.square(
       dimension: size,
       child: CustomPaint(
-        painter: _ArrowPainter(color, orientation, fromCoord, toCoord),
+        painter: (shape.dest == null || shape.orig == shape.dest) ?
```

Comment:
I'd rather have a `switch` logic instead.

```dart
painter: switch (shape) {
   case: Arrow(color: final c, orig: final o, dest: final d): _ArrowPainter(...)
   ...
}
```

It is now possible with dart 3 switch expression and sealed class.